### PR TITLE
Remove edge case that creates issues with redirecting to itself.

### PR DESCRIPTION
**Pull request by dereuromark**
After https://github.com/cakephp/cakephp/issues/6705 I played around with returning all parent calls.
This opened an issue with testing controllers and redirect(null).
Tests that worked before suddenly break when using `return parent::...()`.

We should remove that edge case as it is perfectly fine to redirect to itself (null => same URL), and currently the location header is not properly set in this case.

### CODE REVIEW

**Comment by dereuromark** [comment]
I removed the requirement for the first param, as essentially you already could pass null, so

    $Controller->redirect(null); // Redirect to itself

was the same as it now is possible as

    $Controller->redirect(); // Redirect to itself

**Comment by dereuromark** [comment]
I had to revert the = null addition as it is not BC and throwing strict errors. Maybe we can do that for 3.1?

File: tests/TestCase/Controller/ControllerTest.php [hunk 1]
```diff
@@ -474,6 +474,22 @@ public function testRedirectByCode($code, $msg)
...
 
     /**
+     * testRedirect method
+     *
+     * @return void
+     */
+    public function testRedirectSameUrl()
+    {
+        $Controller = new Controller(null);
+        $Controller->response = new Response();
```
Comment (markstory) [inline tests/TestCase/Controller/ControllerTest.php]: Why not pass the response as a constructor arg?

File: tests/TestCase/Controller/ControllerTest.php [hunk 2]
```diff
@@ -474,6 +474,22 @@ public function testRedirectByCode($code, $msg)
...
+
+        $response = $Controller->redirect(null);
```
Comment (markstory) [inline tests/TestCase/Controller/ControllerTest.php]: I don't think the explicit null is necessarily bad or something we _need_ to fix, but I can see why allowing `redirect()` would be nice as it mirrors `Router::url()`.

File: tests/TestCase/Controller/ControllerTest.php (hunk 1 above)
Comment (dereuromark) [inline tests/TestCase/Controller/ControllerTest.php]: Was a copy and paste from above.

**Comment by lorenzo** [comment]
Not sure I like that "null" means "here"

**Comment by dereuromark** [comment]
@lorenzo As  @markstory pointed out, that is already the case using null explicitly, as Router::url() behaves this way and does not require a first argument either. So it would only be consistent. But that would be something for 3.1.

**Comment by lorenzo** [comment]
I would rather remove that out of the router than make this "consistent"

**Comment by dereuromark** [comment]
If you call redirect() it should not only set the status code as it currently does, but also actually do the redirect. if that happens to be the same URL that sounds totally fine to me. Currently it is broken as it sets 302 and no Location header.

**Comment by markstory** [comment]
@lorenzo What should the short form syntax be for 'here'?

**Comment by lorenzo** [comment]
Do we need array syntax for that? `redirect($this->request->here)` is very clear and easy to understand

**Comment by markstory** [comment]
Very good point @lorenzo. @dereuromark we could raise an exception on empty url instead?

**Comment by ADmad** [comment]
> we could raise an exception on empty url instead?

That won't be BC.

**Comment by markstory** [comment]
@ADmad True, I didn' think of that.

**Comment by ADmad** [comment]
So basically Router::url() accepting null cannot be removed until 4.0

**Comment by josegonzalez** [comment]
Can we log an error at least?

**Comment by ADmad** [comment]
It's not an error, it's expected behavior in current code.

**Comment by josegonzalez** [comment]
CakeLog::warning() ;)

**Comment by dereuromark** [comment]
I dont see an issue with just making it conform to router:: url().
It doesnt hurt, and is BC.

**Comment by ADmad** [comment]
@josegonzalez That would unnecessarily fill up the logs :)

> I dont see an issue with just making it conform to router:: url().

> So it is more consistent, doesn't hurt, and is BC. in 3.1. we can then remove the explicit null call.

@dereuromark I agree.

**Comment by dereuromark** [comment]
Haven't we had strict errors before? Those are technically the same as deprecation warnings, they dont affect productive code.

**Comment by dereuromark** [comment]
wordpress has more E_STRICT errors than lines of code ;) And it also seems to still function :trollface:

**Comment by markstory** [comment]
@dereuromark We currently have 0 strict errors and it would be nice to keep it that way.

**Comment by ADmad** [comment]
While `null` meaning `here` might not be ideal, it isn't that big a deal IMHO. Let's just stick with what we have in 3.x and we can change it in 4.0.

**Comment by lorenzo** [comment]
@ADmad sticking with what we have now means this won't be needed?

**Comment by ADmad** [comment]
@lorenzo Sorry for not being clear, I was referring to `Router::url()`. I am :+1: for this change for consistency with `Router::url()`.

**Comment by lorenzo** [comment]
ok, I'm :-1: I really think it is a bad idea to make the situation even worse

**Comment by ADmad** [comment]
Apart from consistency I don't really care, so I am more like a 0 on this :)

**Comment by dereuromark** [comment]
Worse? The current situation is that it is broken. So this only makes it better as it is not broken afterwards.

**Comment by lorenzo** [comment]
how is it broken now? It just does not allow to pass null

**Comment by dereuromark** [comment]
That it **supports** the explicit null, which then breaks:
```php
$url = $fooBarBaz ?: null;
// We call redirect() so we definitely expect it to redirect
return $this->redirect($url);
```

The result is a broken redirect: 302 status code is set, but no redirect Location header.
That's bad. One in this case expects it to redirect to itself, as no other explicit URL value has been passed. The 302 status code itself is useless and non-functional. The code just silently dies (white blank screen).,

**Comment by lorenzo** [comment]
It is broken if you supply broken values, what about

```
$url = $foo ? : $this->request->here;
```

**Comment by lorenzo** [comment]
to fix the blank screen ww could throw an exception in redirect to prevent the user from passing an empty value

**Comment by dereuromark** [comment]
Null is not a broken value for me, it is a totally acceptable one as long as Router::url() also accepts null to default to itself. That is logic and consistent to me.

**Comment by AD7six** [comment]
> @param string|array $url

Url is a mandatory arg, and null isn't a documented acceptable value. Passing invalid values into a method doesn't make it broken.

**Comment by dereuromark** [comment]
Then feel free to change it to throw an exception in that case. Sounds like the consensus here.
